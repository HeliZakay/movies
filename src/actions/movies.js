import uuid from "uuid";
import database from "../firebase/firebase";
import moment from "moment";


// ADD_MOVIE
export const addMovie = (movie = {
        personName: "",
        movieName: "",
        content: "",
        score: 0,
        createdAt: 0
}) => {
    return ({
      type: 'ADD_MOVIE',
      movie: {
        ...movie,
        id: uuid()
      }
    });
  };


export const startAddMovie = (movieData = {}) => {
  return (dispatch) => {
    const {
      movieName="",
      personName="",
      score=0,
      content="",
      createdAt="0"      
     } = movieData;
     const movie = {movieName, personName, score, content, createdAt: moment(createdAt).format() };
     return database.ref("movies").push(movie).then( (ref) => {
        dispatch(addMovie({
          id: ref.key,
          ...movie
        }));
     });
  };
};

// EDIT_MOVIE
export const editMovie = (id, updates) => ({
  type: 'EDIT_MOVIE',
  id,
  updates
});


  // REMOVE_MOVIE
  export const removeMovie = ({ id } = {}) => ({
    type: 'REMOVE_MOVIE',
    id
  });
  