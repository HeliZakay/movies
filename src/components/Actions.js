import React from "react";
import { Link} from 'react-router-dom';

export default () => {
    return (
        <div className="actions-section">
        <div className="content-container">
        <div className="actions-section__wrapper">
        
        <div className="actions-section__item">
            <Link className="actions-section__link" to="/create">
            <div className="actions-section__background">
            <img className="actions-section__img" src="images/popcorn.png"></img> 
            </div>
            <h3 className="actions-section__text">Add a movie</h3>
            
            </Link>
        </div>
        <div className="actions-section__item">
            <Link className="actions-section__link" to="/friends">
            <div className="actions-section__background">
            <img className="actions-section__img" src="images/familyguy.png"></img>  
            </div>
            <h3 className="actions-section__text">Add a friend</h3>
           
            </Link>
        </div>
        <div className="actions-section__item">
            <Link className="actions-section__link" to="/contact">
            <div className="actions-section__background">
            <img className="actions-section__img" src="images/envelop.png"></img>  
            </div>
            <h3 className="actions-section__text">Send feedback</h3>
            
            </Link>
        </div>
        </div>  
        </div>          
        </div>
    );
};



