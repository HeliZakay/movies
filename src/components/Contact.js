import React from "react";
import {connect} from "react-redux";

export class Contact extends React.Component {
    onSubmit = (event) => {
        event.preventDefault();
        this.props.history.push("/homepage");
    }
    render() {
        return (
            <form
            action="/" 
            method="post" 
            encType="text/plain"
            // onSubmit={this.onSubmit}
            className="form" >

            <div className="page-header">
            <div className="content-container--form">
                 <h2 className="page-header__title">Send Feedback</h2>
            </div>
        </div>
        <div className="content-container--form">
           <p>
            The goal here is to make your experience as much pleasant as possible!
           </p>
           <textarea className="textarea text-area--contact"
                        placeholder="Write your feedback here"
                        name="feedback" autoFocus>
        </textarea>
        <div>
                <button type="submit" className="button button--form"><strong>Send!</strong></button>
        </div>
        </div>
        </form>

        );
    };
}

    
 export default connect()(Contact);
  
