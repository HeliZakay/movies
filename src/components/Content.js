import React from 'react';
import {connect} from "react-redux";

export  class Content extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            readMore: false
        }
    }
    readMore = () => {
        this.setState({readMore: !this.state.readMore});
    }
    shortenContent = (content) => {
        if(content.length< 150) {
            return <p>"{content}"</p>;
        }
        else {
            return <div>
            {!this.state.readMore ?
            <p>
            "{content.slice(0,150)}..."
            <a onClick={this.readMore}>
            {this.props.language === "English"?  "Read More" : "קרא עוד "}
            </a>
            </p>:
            <p>
            "{content}"
            <a onClick={this.readMore}>
            {this.props.language === "English"?  "Read Less" : "קרא פחות "}
            </a>
            </p>
            }
            </div>;
        }
    }
    render() {
        return (
            <div className="content">
             {this.props.content && <div className="card-text"> {this.shortenContent(this.props.content)}</div>}           
            </div>
        );
    }
};



const mapStateToProps = (state) => ({
    language: state.auth.language
});

export default connect(mapStateToProps)(Content);
   