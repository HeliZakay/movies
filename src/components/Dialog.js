import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import MuiDialogTitle from '@material-ui/core/DialogTitle';
import MuiDialogContent from '@material-ui/core/DialogContent';
import MuiDialogActions from '@material-ui/core/DialogActions';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';
import Typography from '@material-ui/core/Typography';
import {connect} from "react-redux";
import MovieCard from "./MovieCard";


const styles = (theme) => ({
  root: {
    margin: 0,
    padding: theme.spacing(2),
  },
  closeButton: {
    position: 'absolute',
    right: theme.spacing(1),
    top: theme.spacing(1),
    color: theme.palette.grey[500],
  }
});

const DialogTitle = withStyles(styles)((props) => {
  const { children, classes, onClose, ...other } = props;
  return (
    <MuiDialogTitle disableTypography className={classes.root} {...other}>
      <Typography variant="h6">{children}</Typography>
      {onClose ? (
        <IconButton aria-label="close" className={classes.closeButton} onClick={onClose}>
          <CloseIcon />
        </IconButton>
      ) : null}
    </MuiDialogTitle>
  );
});

const DialogContent = withStyles((theme) => ({
  root: {
    padding: theme.spacing(2),
  },
}))(MuiDialogContent);

export function CustomizedDialogs(props) {
  const [open, setOpen] = React.useState(false);

  const handleClickOpen = () => {
    setOpen(true);
  };
  const handleClose = () => {
    setOpen(false);
  };

  return (
    <div >
    <div className="dialog__button-div">
      <button className="btn button-friend--message btn-lg button-open-review" onClick={handleClickOpen}>
       {props.language === "English"? "Open review": "פתח ביקורת"}
      </button>
      </div>
      <Dialog 
      className={String(props.language !== "English" && "align-right")}
      onClose={handleClose} aria-labelledby="customized-dialog-title" open={open}>
        <DialogTitle id="customized-dialog-title" onClose={handleClose}>
          {props.language === "English" || !props.movie.hname?
          (props.movie.movieName ): (props.movie.hname)}
        </DialogTitle>
        <DialogContent dividers>
         
           <MovieCard 
                        className={String(props.language !== "English" && "align-right")}
                        reviews = {props.movie.reviews}
                        id={props.movie.id}
                        movieName={props.movie.movieName}
                        hname={props.movie.hname}
                        imdbData={props.movie.imdbData}
                        dialog={true}
                     />
        </DialogContent> 
      </Dialog>
    </div>
  );
}

const mapStateToProps = (state) => ({
  language: state.auth.language,
  movies: state.movies
});
export default connect(mapStateToProps)(CustomizedDialogs);