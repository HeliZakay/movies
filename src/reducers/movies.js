// Movie Reducer


export default (state = [], action) => {
    switch (action.type) {
      case 'ADD_MOVIE':
        return [
          ...state,
          action.movie
        ];
        case 'EDIT_MOVIE':
          return state.map((movie) => {
            if (movie.id === action.movieId) {
              const updatedReviewsArray =  movie.reviews.map((review) => {
                if (review.id === action.reviewId) {
                  return {
                    ...review,
                    ...action.updatedReview
                  };
                }
                else {
                  return review;
                };
              });
              return {
                ...movie,
                reviews: updatedReviewsArray
              }
            } else {
              return movie;
            };
          });
      case 'REMOVE_MOVIE':
        return state.map((movie) => {
          if (movie.id === action.movieId){
           return {
              ...movie,
              reviews: movie.reviews.filter((review) => review.id !== action.reviewId)
            };
          } else {
            return movie;
          }
        });
      case "REMOVE_MOVIE_COMPLETELY":
        return state.filter((movie) => {
          return movie.id !== action.movieId;
        });
      case "SET_MOVIES":
        return action.movies;
      case "ADD_REVIEW":
          return state.map((movie) => {
            if (movie.id === action.movieId) {
              const updatedReviewsArray = [action.review, ...movie.reviews];
                return {
                  ...movie, 
                  reviews: updatedReviewsArray
              }
            }
            else {return movie};
          });
      case "GIVE_STAR_TO_REVIEW":
        return state.map((movie) => {
          if(movie.id === action.movieId) {
            const updtaedReviews = movie.reviews.map((review) => {
              if (review.id === action.reviewId) {
                let updatedStarsArray;
                if (review.stars) {
                  updatedStarsArray = [...review.stars, action.user];
                } else {
                  updatedStarsArray = [action.user];
                }
                return {
                  ...review,
                  stars: updatedStarsArray
                }
              } else {
                return review;
              }
            });
            return {
              ...movie,
              reviews: updtaedReviews
            }
          } else {
            return movie;
          }
        })
      default:
        return state;
    }
  };
