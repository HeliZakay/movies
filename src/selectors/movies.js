import FriendDialog from "../components/FriendDialog";

  export const getHname = (movies, movieName) => {
    const movie = movies.filter((movie) => {
      return movie.movieName === movieName;
    });
    if (movie && movie[0] && movie[0].hname) {
      return movie[0].hname;
    }
    else {
      return undefined;
    }
  }
  
  
  export const didIGaveStarToReview = (stars, uid) => {
    let result = false;
    if (stars) {
      stars.forEach((star) => {
        if (star.uid === uid) {
          result = true;
        }
      });
    }
    return result;
  }
  
  export const sortReviewsByDateAndPerson = (reviews, person) => { 
    if (!person) {
      return reviews.sort((a,b) => { 
        return a.createdAt < b.createdAt? 1:-1;
      });
    } else {
      const sortedByDate = reviews.sort((a,b) => { 
        return a.createdAt < b.createdAt? 1:-1;
      });
     const personReview =  reviews.filter((review) => {
        return review.personName.toLowerCase().includes(person.toLowerCase());
      });
      return [...personReview, ...sortedByDate.filter((review)=> {
        return !review.personName.toLowerCase().includes(person.toLowerCase());
      })];
    }   
  };

  
const computeAverageScore = (movie) => {
  let sum = 0;
  movie.reviews.forEach((review) => {
      sum += Number(review.score);
  });
  const result = sum/movie.reviews.length;
  return  Math.round(result*2)/2;;
}

// Get visible movies
export default (movies, { text, sortBy, person, genres}, language) => {
    return movies.filter((movie) => {
      let textMatch;
      if (language === "English") {
        textMatch = movie.movieName.toLowerCase().includes(text.toLowerCase());
      } else {
        textMatch = String(movie.hname).includes(text);
      }
      let personMatch = false;
      movie.reviews.forEach((review) => {
        if (review.personName.toLowerCase().includes(person.toLowerCase())) {
          personMatch = true;
        }
      });
      let genreMatch = false;
      if (!genres) {
        genreMatch = true;
      } else if (genres.length ===0) {
        genreMatch = true;
      } else if (!movie.imdbData.Genre) {
        genreMatch = false;
      } else {
        const genresArray = movie.imdbData.Genre.split(", ");
        genresArray.forEach((genre) => {
          if (genres.includes(genre)) {
            genreMatch = true;
          }
        });
      }
      return genreMatch && textMatch && personMatch;
    }).map((movie) => {
      const sorted = sortReviewsByDateAndPerson(movie.reviews, person);
      return {...movie, reviews: sorted};
    }).sort((a, b) => {
      if (sortBy === 'date') {
        return a.reviews[0].createdAt < b.reviews[0].createdAt ? 1 : -1;
      } else if (sortBy === 'score') {
        return computeAverageScore(a) < computeAverageScore(b) ? 1 : -1;
      }
    });
  };

  export const getMovieById = (movies, movieId) => {
    let result = undefined;
    movies.forEach((movie) => {
      if(movie.id === movieId) {
        result = movie;
      }
    });
    return result;
  };

  export const getMovieReviews = (movieId, movies) => {
    let reviews = [];
    movies.forEach((movie) => {
      if (movie.id === movieId) {
        reviews=movie.reviews;
      }
    });
    return reviews;
  }

  export const isMovieNameExistsAlready = (movieName, movies) => {
    let result = undefined;
    movies.forEach((movie) => {
      if (movie.movieName === movieName) {
        result=movie.id;
      }
    });
    return result;
  }
export const findMovie = (movies, movieName, personName, uid, type) => {
  const movieArr =  movies.filter((movie) => {
    return movie.movieName === movieName;
  });  
  if(movieArr && movieArr[0]) {
    const movie = movieArr[0];
    const oldReviews = movie.reviews;
    const filteredReviews = oldReviews.filter((review) => {
      if(type==="newReview") {
        return review.personName !== personName;
      }
      else {
        return review.userUid !== uid;
      }
    });
    const firstReview = oldReviews.filter((review) => {
      if(type==="newReview") {
        return review.personName === personName;
      }
      else {
        return review.userUid === uid;
      }
    });
    const sortedReviews= [...firstReview, ...filteredReviews];
    return {
      ...movie,
      reviews: sortedReviews
    };
  } else {
    return undefined;
  }
  
}

export const getUnseenMovies = (movies, uid, watchlist) => {
  return movies.filter((movie)=> {
    const reviewsUids = movie.reviews.map((review)=> {
      return review.userUid;
    });
    return (!reviewsUids.includes(uid) && !watchlist.includes(movie.id) && computeAverageScore(movie) >=7);
  });
}

export const takeTop = (array, num) => {
  if (array.length <=num) {
    return array;
  }
  else {
    return array.slice(0,num);
  }
}

export const didFriendReviewedMovie = (reviews, friendId) => {
  
  const result = reviews.filter((review) => {
    return review.userUid === friendId;
  });
  return result.length > 0;

}

const didSendReviewToMovie = ({reviews, uid}) => {
  let reviewId;
  reviews.forEach((review) => {
    if (review.userUid === uid) {
      reviewId = review.id;
    }
  });
  return reviewId;
}

export const getMyMovies = (movies, uid) => {
  return movies.filter((movie) => {
    const myReviewId = didSendReviewToMovie({reviews: movie.reviews, uid});
    if (myReviewId) {
      let hightRatedByMe = false;
      movie.reviews.forEach((review) => {
        if(review.id === myReviewId && review.score >=8){
          hightRatedByMe = true;
        }
      });
      return hightRatedByMe;
    } else {
      return false;
    }
  });
} 