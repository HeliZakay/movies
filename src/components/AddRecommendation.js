import React from "react";
import RecommendationForm from "./RecommendationForm";
import {connect} from "react-redux";
import {startAddMovie} from "../actions/movies";

export class AddRecommendation extends React.Component {
   onSubmit = (movie) => {
      this.props.startAddMovie(movie);
      this.props.history.push("/");
   };
   render() {
      return (
         <div>
         <h1>Add Recommendation</h1>
         <RecommendationForm
           onSubmit={this.onSubmit}
          />
         </div>
      );
   }
}


const mapDispatchToProps = (dispatch) => ({
   startAddMovie: (movie) => dispatch(startAddMovie(movie))
});
   
export default connect(undefined, mapDispatchToProps)(AddRecommendation);

// import React from 'react';
// import { connect } from 'react-redux';
// import RecommendationForm from './RecommendationForm';
// import { startAddMovie } from '../actions/movies';

// export class AddMovie extends React.Component {
//   onSubmit = (movie) => {
//     this.props.startAddMovie(movie);
//     this.props.history.push('/');
//   };
//   render() {
//     return (
//       <div>
//         <h1>Add Movie</h1>
//         <RecommendationForm
//           onSubmit={this.onSubmit}
//         />
//       </div>
//     );
//   }
// }

// const mapDispatchToProps = (dispatch) => ({
//   startAddMovie: (movie) => dispatch(startAddMovie(movie))
// });

// export default connect(undefined, mapDispatchToProps)(AddMovie);