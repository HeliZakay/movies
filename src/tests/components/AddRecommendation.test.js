import React from "react";
import {shallow} from "enzyme";
import {AddRecommendation} from "../../components/AddRecommendation";
import movies from "../fixtures/movies";

let startAddMovie, history, wrapper;

beforeEach(() => {
    startAddMovie = jest.fn();
    history = { push: jest.fn() };
    wrapper =shallow(<AddRecommendation startAddMovie={startAddMovie} history={history} />);
});

test("should render AddRecommendation correctly", () => {
    expect(wrapper).toMatchSnapshot();
});

test("should handle onSubmit", () => {
    wrapper.find("RecommendationForm").prop("onSubmit")(movies[1]);
    expect(history.push).toHaveBeenLastCalledWith("/");
    expect(startAddMovie).toHaveBeenLastCalledWith(movies[1]);
});