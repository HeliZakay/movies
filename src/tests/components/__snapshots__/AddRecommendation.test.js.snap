// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddRecommendation correctly 1`] = `
<div>
  <h1>
    Add Recommendation
  </h1>
  <RecommendationForm
    onSubmit={[Function]}
  />
</div>
`;
